### PR TITLE
Produce .version file for current project

### PR DESCRIPTION
**Pull request by blender**
This PR addresses #2634 

~~The core method assumes that __the last path component of the current directory is the the name of the current project__ as it would appear in the Cartfile.~~

The core method derives the name of the current project as it would appear in the Cartfile. form `git remote get-url origin`

I'm open to feedback and to discuss other ideas.

### CODE REVIEW

**Comment by blender** [comment]
I didn't add test. Any ideas on how to test this?

File: Source/CarthageKit/VersionFile.swift [hunk 1]
```diff
@@ -237,6 +237,40 @@ struct VersionFile: Codable {
...
+) -> SignalProducer<(), CarthageError> {
+	let currentGitTagOrCommitish = launchGitTask(["rev-parse", "HEAD"])
+		.map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
+		.flatMap(.merge) { headCommitish in
+			return launchGitTask(["describe", "--tags", "--exact-match", headCommitish])
+				.map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
+				.flatMapError { _  in SignalProducer(value: headCommitish) }
+		}
+
+	let dependecyName = !rootDirectoryURL.lastPathComponent.isEmpty ? rootDirectoryURL.lastPathComponent : "Current-Unknown"
```
Comment (mdiep) [inline Source/CarthageKit/VersionFile.swift]: _dependency_ is misspelled here

**Comment by mdiep** [comment]
> The core method assumes that **the last path component of the current directory is the the name of the current project** as it would appear in the Cartfile.

Would it make more sense to make this as `.version` or `._.version`?

**Comment by blender** [comment]
The idea is to generate the file compatible with `--cache-builds` with no extra steps. `--cache-builds` assumes `<name-as-in-cartfile>.version`. 

`.version` or `._.version` would be a hidden files by default and I can already see the complaints incoming. 
I didn't go with  current `Current.version` because... I can already see someone having a dependency called `Current`.

I though that defaulting to the name of the current directory was less or a wild guess.

I am open to change it to a predictable name so that it can be renamed with a script in simple steps.

**Comment by mdiep** [comment]
I think it'd be better _not_ to use the name of the current directory. That way you can copy the build directory around without it depending on the last component of the project directory. There are lots of reasons you might want to do that.

**Comment by blender** [comment]
How about using the name of the current project or workspace? So

```
.<projectOrWorkspaceName>.version
```

**Comment by mdiep** [comment]
That seems like it could lead to conflicts. 🤔 

Maybe we could use the primary remote (if it exists) and calculate the name just like we'd do for a dependency? Falling back to something like what you have now?

**Comment by blender** [comment]
I’ll investigate and ping back.

**Comment by blender** [comment]
@mdiep relevant bit: https://github.com/Carthage/Carthage/pull/2636/commits/5627347d38eb6b15ac917b37bf2dd1f16e2a3379#diff-6206d1bf8ed3195b3b646bc3d2b6553dR253

I assumed 1 remote (origin) is enough. If we get complaints we can check all remotes and pick the most common name.

Also squashed commits

**Comment by blender** [comment]
@mdiep ping

File: Source/CarthageKit/VersionFile.swift [hunk 2]
```diff
@@ -237,6 +237,43 @@ struct VersionFile: Codable {
...
+///
+/// Returns a signal that succeeds once the file has been created.
+public func createVersionFileForCurrentProject(
+	platforms: Set<Platform>,
+	buildProducts: [URL],
+	rootDirectoryURL: URL
+) -> SignalProducer<(), CarthageError> {
+	let currentProjectAsDependecy = launchGitTask(["remote", "get-url", "origin"])
+		.map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
+		.map { Dependency.git(GitURL($0)) }
```
Comment (mdiep) [inline Source/CarthageKit/VersionFile.swift]: There's not always a remote named `origin`, so this should probably catch errors and default to another value.

File: Source/CarthageKit/Xcode.swift [hunk 3]
```diff
@@ -818,15 +825,24 @@ public func buildInDirectory( // swiftlint:disable:this function_body_length
...
-				.flatMapError { _ in .empty }
+				if isCurrentProject {
+					return createVersionFileForCurrentProject(platforms: options.platforms,
+							buildProducts: urls,
+							rootDirectoryURL: rootDirectoryURL
+						)
+						.flatMapError { _ in .empty }
+				} else {
+					return createVersionFile(
+						for: dependency!.dependency,
```
Comment (mdiep) [inline Source/CarthageKit/Xcode.swift]: Instead of checking `== nil` above and then using `!` here, I'd prefer to use `if let` or `guard let`.

**Comment by mdiep** [comment]
Looks good overall. 👍 Just a couple things to clean up.

File: Source/CarthageKit/VersionFile.swift (hunk 2 above)
Comment (blender) [reply Source/CarthageKit/VersionFile.swift]: ok so I will go the full way.

Comment (mdiep) [reply Source/CarthageKit/VersionFile.swift]: Otherwise you could fallback to a string like "CurrentProject" or something

**Comment by blender** [comment]
So yeah, it might have gotten a little out of hand. https://github.com/Carthage/Carthage/pull/2636/commits/3d80665891950f582f7fa6e307ced88a99d111c0

Anyways I am not fully happy because I would also notify the user that we're defaulting to `_Current.version` (which I picked semi arbitrarily). However this requires some rethinking of the logging.

**Comment by mdiep** [comment]
I'm a little behind from Thanksgiving last week, but I'll try to review this tomorrow or at least sometime this week.

File: Source/CarthageKit/VersionFile.swift [hunk 4]
```diff
@@ -237,6 +237,72 @@ struct VersionFile: Codable {
...
+	// List all remotes known for this repository
+	let allRemotes = launchGitTask(["remote"])
+		.flatMap(.concat) { $0.linesProducer }
+		.map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
+
+	// For each remote, get the `get-url`
+	let allRemoteURLs = allRemotes
+		.flatMap(.concat) { launchGitTask(["remote", "get-url", $0]) }
+		.map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
+		.collect()
```
Comment (mdiep) [inline Source/CarthageKit/VersionFile.swift]: You could probably use `git remote -v` to get the URLs instead.

File: Source/CarthageKit/VersionFile.swift [hunk 5]
```diff
@@ -237,6 +237,72 @@ struct VersionFile: Codable {
...
+		.map { (remoteURLPopularityMap: [String: Int]) -> (url: String, popularity: Int) in
+			remoteURLPopularityMap.sorted { lhs, rhs in
+				if lhs.value == rhs.value {
+					return lhs.key < rhs.key
+				}
+				return lhs.value > rhs.value
+			}
+		    // This should never happen... but use a fallback anyways
+			.first ?? ("_Current", 0)
+		}
```
Comment (mdiep) [inline Source/CarthageKit/VersionFile.swift]: I think it'd be nice to default to `origin` if it exists.

* Most people will only have 1 remote.
* Most people with >1 remote will have an `origin`.
* If you don't have either, who knows what you'll have?

But choosing `origin` over the "popular" or alphabetical option seems like a better choice.

File: Source/CarthageKit/VersionFile.swift [hunk 6]
```diff
@@ -237,6 +237,72 @@ struct VersionFile: Codable {
...
+		// Sort remotes by popularity or alphabetically in case of a draw
+		.map { (remoteURLPopularityMap: [String: Int]) -> (url: String, popularity: Int) in
+			remoteURLPopularityMap.sorted { lhs, rhs in
+				if lhs.value == rhs.value {
+					return lhs.key < rhs.key
+				}
+				return lhs.value > rhs.value
+			}
+		    // This should never happen... but use a fallback anyways
```
Comment (mdiep) [inline Source/CarthageKit/VersionFile.swift]: Probably someone will try doing this before pushing their repo anywhere.

File: Source/CarthageKit/VersionFile.swift (hunk 4 above)
Comment (blender) [reply Source/CarthageKit/VersionFile.swift]: ah, nice. definitely.

File: Source/CarthageKit/VersionFile.swift (hunk 5 above)
Comment (blender) [reply Source/CarthageKit/VersionFile.swift]: ok

File: Source/CarthageKit/VersionFile.swift (hunk 6 above)
Comment (blender) [reply Source/CarthageKit/VersionFile.swift]: You bet. I will explain that. Should I keep `_Current` as fallback from `origin` ?

File: Source/CarthageKit/Xcode.swift (hunk 3 above)
Comment (blender) [reply Source/CarthageKit/Xcode.swift]: Yes, sorry. I missed it.

File: Source/CarthageKit/VersionFile.swift (hunk 4 above)
Comment (blender) [reply Source/CarthageKit/VersionFile.swift]: Done

File: Source/CarthageKit/VersionFile.swift (hunk 5 above)
Comment (blender) [reply Source/CarthageKit/VersionFile.swift]: Done

**Comment by blender** [comment]
@mdiep  I have made the changes. I have tests with:
- malformed remotes ✅
- no remotes ✅

File: Source/CarthageKit/VersionFile.swift [hunk 7]
```diff
@@ -239,6 +239,101 @@ struct VersionFile: Codable {
...
+/// Derives the current project name from `git remote get-url origin`
+///
+/// Returns a signal that succeeds once the file has been created.
+public func createVersionFileForCurrentProject(
+	platforms: Set<Platform>,
+	buildProducts: [URL],
+	rootDirectoryURL: URL
+) -> SignalProducer<(), CarthageError> {
+
+	/*
```
Comment (blender) [inline Source/CarthageKit/VersionFile.swift]: @mdiep cases I have covered

File: Source/CarthageKit/VersionFile.swift [hunk 8]
```diff
@@ -239,6 +239,101 @@ struct VersionFile: Codable {
...
+			guard let origin = remotePopularityMap["origin"] else {
+				let urlOfMostPopularRemote = remotePopularityMap.sorted { lhs, rhs in
+					if lhs.value.popularity == rhs.value.popularity {
+						return lhs.key < rhs.key
+					}
+					return lhs.value.popularity > rhs.value.popularity
+				}
+				.first?.value.remoteNameAndURL.url
+
+				// If the reposiroty is not pushed to any remote
```
Comment (blender) [inline Source/CarthageKit/VersionFile.swift]: @mdiep  Not a pushed repo like we discussed

**Comment by mdiep** [comment]
Looks good! Thanks for all your work on this one!

**Pull request merged**